internal/graph/check_test.go

```diff
@@ -731,6 +741,41 @@ type document
 			minDBReads: 2, // one read to find org:fga + (one direct check) to see if user:jon is a member of org:fga
 			maxDBReads: 2,
 		},
+		{
+			name:       "userset_no_access_1",
+			check:      tuple.NewTupleKey("document:no_access", "userset", "user:maria"),
+			allowed:    false,
+			minDBReads: 1, // 1 userset read (none found)
+			maxDBReads: 1,
+		},
+		{
+			name:       "userset_no_access_2",
+			check:      tuple.NewTupleKey("document:x", "userset", "user:no_access"),
+			allowed:    false,
+			minDBReads: 2, // 1 userset read (1 found) follow by 1 direct tuple check (not found)
+			maxDBReads: 2,
+		},
+		{
+			name:       "userset_access",
+			check:      tuple.NewTupleKey("document:x", "userset", "user:maria"),
+			allowed:    true,
+			minDBReads: 2, // 1 userset read (1 found) follow by 1 direct tuple check (found)
+			maxDBReads: 2,
+		},
+		{
+			name:       "multiple_userset_no_access",
+			check:      tuple.NewTupleKey("document:x", "multiple_userset", "user:no_access"),
+			allowed:    false,
+			minDBReads: 3, // 1 userset read (2 found) follow by 2 direct tuple check (not found)
+			maxDBReads: 3,
+		},
+		{
+			name:       "multiple_userset_access",
+			check:      tuple.NewTupleKey("document:x", "multiple_userset", "user:maria"),
+			allowed:    true,
+			minDBReads: 2, // 1 userset read (2 found) follow by 1 direct tuple check (found, returns immediately)
+			maxDBReads: 2,
```

Comment:
This should be 3. In the worst case the read of the second userset branches occurs first. There's a race on the userset evaluations.

The resolved branch will require 2 database queries, but the max database reads is 3.

Reply:
we may execute 3 reads overall but we will only return 2 because we return the count in the path where we found `allowed=true`, which is not going to be aware of the read happening in the other goroutine

Reply:
Right - we agree. But this test assumes the max db reads is the upper bound of how many total may get executed. That's my point.

Reply:
no, this test asserts on the response field's datastore count, which will never be 3. I'd get test failures otherwise

Reply:
What's the purpose of the `maxDBReads` anymore then? In this example it's possible at most 3 DB reads will occur - 1 for the initial userset read, and 2 direct queries for each of the usersets found. If we're only ever returning the "resolved" DB query count, then what's the purpose of defining the `maxDBReads`? The number of DB queries required to resolve is different than the total number of DB queries invoked per request, so that's also not clear to me here.

Could we clarify that?